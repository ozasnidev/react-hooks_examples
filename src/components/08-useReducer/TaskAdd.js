import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TaskAdd = ({ handleAddTask }) => {
    const [{ description }, handleInputChange, resetForm] = useForm({
        description: ''
    });
    const handleSubmit = (event) => {
        event.preventDefault();

        if(description.trim().length >= 1){
            handleAddTask({type: 'add', payload: {id: new Date().getTime(), desc: description, done: false}});
            resetForm();
        }
    };

    return (
        <>
            <div className="col-5">
                <h3>Agregar Task</h3>
                <hr/>
                <form
                    onSubmit={handleSubmit}
                >
                    <input
                        type="text"
                        name="description"
                        className="form-control"
                        placeholder="Task"
                        autoComplete="off"
                        value={ description }
                        onChange={ handleInputChange }
                    />
                    <button
                        type="submit"
                        className="btn btn-outline-success mt-2 btn-block"
                    >
                        Agregar
                    </button>
                </form>
            </div>
        </>
    )
}
