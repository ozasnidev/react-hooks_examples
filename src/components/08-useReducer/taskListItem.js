import React from 'react'

export const TaskListItem = ({task, index, handleDeleteTask, handleToggle}) => {
    return (
        <li
            key={task.id}
            className="list-group-item list-group-item-warning m-1"
        >
            <p  
                className={`${task.done && 'complete'}`}
                onClick={() => handleToggle(task.id)}
            >
                {`${index + 1} ${task.desc}`}
            </p>
            <button 
                className="btn btn-outline-danger"
                onClick={() => handleDeleteTask(task.id)}
            >
                Borrar
            </button>
        </li>
    )
}
