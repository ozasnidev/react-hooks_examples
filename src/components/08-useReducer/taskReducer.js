export const taskReducer = ( state = [], action ) => {
    switch(action.type){


        default: return state;
    }
} 