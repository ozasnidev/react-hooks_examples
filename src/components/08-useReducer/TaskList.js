import React from 'react'
import { TaskListItem } from './taskListItem'

export const TaskList = ({taskState, handleDeleteTask, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
            {
                taskState.map((task, index  ) => (
                    <TaskListItem 
                        key = {task.id}
                        task = {task}   
                        index = {index}
                        handleDeleteTask = {handleDeleteTask}
                        handleToggle = {handleToggle}
                    />
                ))
            }
        </ul>
    )
}
