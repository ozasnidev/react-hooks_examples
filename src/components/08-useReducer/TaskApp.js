import React, { useEffect, useReducer } from 'react'
import { taskReducer } from './taskReducer';
import { TaskList } from './TaskList';
import { TaskAdd } from './TaskAdd';
import './styles.css';

const init = () => {
    return JSON.parse(localStorage.getItem('tasks')) || [];
}

export const TaskApp = () => {
    const [taskState, dispatch] = useReducer(taskReducer, [], init);
    const handleDeleteTask = (taskId) => { dispatch({ type: 'delete', payload: taskId}); };
    const handleToggle = (taskId) => { dispatch({type: 'toggle', payload: taskId}); }
    const handleAddTask = (newTask) => { dispatch(newTask); }
    useEffect( () => {
        localStorage.setItem('tasks', JSON.stringify(taskState));
    }, [taskState]);

    return (
        <>
            <h1>TaskApp ({taskState.length})</h1>
            <hr/>
            <div className="row">
                <div className="col-7">
                    <h3>Task</h3>
                    <hr/>
                    <TaskList 
                        taskState = {taskState}
                        handleDeleteTask = {handleDeleteTask}
                        handleToggle = {handleToggle}
                    />
                </div>
                <TaskAdd
                    handleAddTask = {handleAddTask}
                />
            </div>
        </>
    )
}
