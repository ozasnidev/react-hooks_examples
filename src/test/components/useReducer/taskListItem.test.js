import React from 'react';
import { shallow } from 'enzyme';
import { TaskListItem } from '../../../components/08-useReducer/taskListItem';
import { mockTask } from '../../fixtures/mockTask';
import '@testing-library/jest-dom';

describe('TaskListItem component test', ()=> {
    const handleDeleteTask = jest.fn();
    const handleToggle = jest.fn();
    const wrapper = shallow(<TaskListItem task={mockTask[0]} index={0} handleDeleteTask = {handleDeleteTask} 
        handleToggle={handleToggle}/>);
    
    test('render', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('handleDeleteTask function call', () => {
        wrapper.find('button').simulate('click');
        expect(handleDeleteTask).toHaveBeenCalledWith(mockTask[0].id);
    });
    test('handleToggle function call', () => {
        wrapper.find('p').simulate('click');
        expect(handleToggle).toHaveBeenCalledWith(mockTask[0].id);
    });
    test('text p succesfull', () => {
        expect(wrapper.find('p').text().trim()).toBe(`1 ${mockTask[0].desc}`);
    });
    test('text p must have complete class if task.done = true', () => {
        const mockTaskObject = mockTask[0];
        mockTaskObject.done = true;
        const wrapper = shallow(<TaskListItem task={mockTaskObject} />);

        expect(wrapper.find('p').hasClass('complete')).toBeTruthy();
    });
});