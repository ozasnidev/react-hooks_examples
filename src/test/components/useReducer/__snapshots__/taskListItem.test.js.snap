// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskListItem component test render 1`] = `
<li
  className="list-group-item list-group-item-warning m-1"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1 Aprender React.js
  </p>
  <button
    className="btn btn-outline-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
